plugins/experimental/icap/icap_plugin.cc

```diff
@@ -0,0 +1,1267 @@
+/** @file
+
+  @brief A plugin that sends the http body through an ICAP request to a
+         scanner server. If malicious content is detected, then the scanner
+         will return an error message body, which we will pass to the user
+         agent. Otherwise it will return the same content that was passed to
+         it, in which case we will pass the content to user agent.
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#include <string>
+#include <regex>
+
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <iostream>
+#include <sstream>
+
+#include "ts/ts.h"
+
+#define PLUGIN_NAME "icap_plugin"
+
+#define STATE_BEGIN 1
+#define STATE_CONNECT 2
+#define STATE_WRITE_HEADER 3
+#define STATE_WRITE_BODY 4
+#define STATE_READ_ICAP_HEADER 5
+#define STATE_READ_HTTP_HEADER 6
+#define STATE_READ_HTTP_BODY 7
+#define STATE_SEND_ERROR_MSG 8
+#define STATE_BYPASS 9
+#define STATE_BUFFER_OS_RESP 10
+#define STATE_SEND_OS_RESP 11
+
+#define ICAP_SERVICE_URL "icap://127.0.0.1/avscan"
+#define ICAP_VERSION "1.0"
+
+struct TransformData {
```

Comment:
Instead of one big struct consider an approach more like this:  https://github.com/ywkaras/trafficserver/blob/edf6db2f30acb2b480dea1356a935c2c53d62a59/tests/gold_tests/pluginTest/tsapi/ts_tcp.h#L546